### PR TITLE
Include description in repository search.

### PR DESCRIPTION
**Pull request by davidsvantesson**
When doing a keyword search for repositories (for example explore page), include repository description in search. The repository name can be something imaginary / made-up or a product number, while the description actually explains what it does. Therefore you want to include the repository description in the search. This PR should solve #7264.

Do you see any case you don't want repository description included in search, for performance issues or other reasons? It could be made a setting in app.ini, but I didn't see any reason for it.

### CODE REVIEW

**Review by guillep2k — APPROVED** [review]
I don't think there should be any performance problems as all the rows must be scanned anyway. The only problem is that partial matches might make this very problematic. For instance, several repos might have in their descriptions something like "utilizes library from repoabc", so if you look for "repoabc" the system will return not only that repo, but all the others that mention it in their description.

Whether this is a good or a bad thing, is a matter of opinion. For me it would be good, but I'd like others to voice their opinion about it.

**Review by lunny — CHANGES_REQUESTED** [review]
I think it will be a performance problem on a big Gitea instance. For a smaller instance, search maybe an unnecessary things.

**Comment by guillep2k** [comment]
> 
> 
> I think it will be a performance problem on a big Gitea instance. For a smaller instance, search maybe an unnecessary things.

I imagine a big single-server system to have... 10,000 repositories? That would be about ~6MB `repository` table, most likely completely cached.

**Comment by davidsvantesson** [comment]
@lunny Can it be made an app.ini setting (default off)? That way today behavior remains when upgrading, for those that don't want it (due to performance or search including too much).

**Review by sapk — CHANGES_REQUESTED** [review]
This could be an option for performance issue. The function should also be rename/refactor to indicate that it doesn't only search by name. Maybe this could be a common func SearchRepositoryBy and two func SearchRepositoryByName SearchRepositoryByNameAndDesc that call this common function depending of settings.

**Comment by davidsvantesson** [comment]
Based on comments, I have done the following:
- Added app.ini setting (SEARCH_REPO_DESCRIPTION = false)
- Added option to API whether to include description not not (default false, not affected by app.ini)
- If enabled, description will be searched on all pages where description is seen on result page: Explore page, Organization and User profile page. Also on admin repo search page (for consistency with the other search pages).
- Description not searched on "incremental" search on dashboard as you only see the repo name, also ajax makes it performance critical (it uses the API).

I also added a small test.

File: models/repo_list_test.go [hunk 1]
```diff
@@ -74,6 +74,34 @@ func TestSearchRepositoryByName(t *testing.T) {
 	assert.Empty(t, repos)
 	assert.Equal(t, int64(0), count)
 
+	// Test search within description
+	repos, count, err = SearchRepository(&SearchRepoOptions{
```
Comment (sapk) [inline models/repo_list_test.go:78]: Please update the test function name to reflect the new name `func TestSearchRepository(t *testing.T)` (even if it still test also SearchRepositoryByName the real func tested is SearchRepository)

**Review by sapk — CHANGES_REQUESTED** [review]
Looking good just one more nit

File: models/repo_list_test.go (hunk 1 above)
Comment (davidsvantesson) [reply models/repo_list_test.go:78]: Fixed

File: custom/conf/app.ini.sample [hunk 2]
```diff
@@ -116,6 +116,8 @@ DEFAULT_THEME = gitea
 THEMES = gitea,arc-green
 ; Whether the full name of the users should be shown where possible. If the full name isn't set, the username will be used.
 DEFAULT_SHOW_FULL_NAME = false
+; Whether to search within description at repository search on explore page.
+SEARCH_REPO_DESCRIPTION = false
```
Comment (guillep2k) [inline custom/conf/app.ini.sample]: I know that I was not the one worried about performance in this case, but I think `SEARCH_REPO_DESCRIPTION` should default to true.  @sapk, @lunny

Comment (lafriks) [reply custom/conf/app.ini.sample]: I also agree that this should default to true

Comment (sapk) [reply custom/conf/app.ini.sample]: ok for me to setting it to true if you want

**Review by sapk — APPROVED** [review]

File: custom/conf/app.ini.sample (hunk 2 above)
Comment (davidsvantesson) [reply custom/conf/app.ini.sample]: I set it to false to not change behavior for existing installations, but I think most will want to have it on. @lunny Your opinion on this?

Comment (lafriks) [reply custom/conf/app.ini.sample]: It does change behavior but we tend to enable new features by default if only they have dramatic change in performance/disk usage. I don't think this have such big impact on performance

Comment (davidsvantesson) [reply custom/conf/app.ini.sample]: Changed to default true!

**Comment by lunny** [comment]
> @lunny Can it be made an app.ini setting (default off)? That way today behavior remains when upgrading, for those that don't want it (due to performance or search including too much).

I think this is a good idea.

**Comment by lafriks** [comment]
@lunny need you approval ;)

**Review by lunny — APPROVED** [review]

**Pull request merged**